lib/ui/text.dart

```diff
@@ -947,6 +947,68 @@ class ParagraphConstraints {
   String toString() => '$runtimeType(width: $width)';
 }
 
+/// Defines various ways to vertically bound the boxes returned by
+/// [Paragraph.getBoxesForRange].
+enum BoxHeightStyle {
+    /// Provide tight bounding boxes that fit heights per run. This style may result
+    /// in uneven bounding boxes that do not nicely connect with the boxes above,
+    /// below, and adjacent.
+    tight,
+
+    /// The height of the boxes will be the maximum height of all runs in the
+    /// line. All boxes in the same line will be the same height. This does not
+    /// guarantee that the boxes will cover the entire vertical height of the line
+    /// when there is additional line spacing.
+    ///
+    /// See [RectHeightStyle.includeLineSpacingTop], [RectHeightStyle.includeLineSpacingMiddle],
+    /// and [RectHeightStyle.includeLineSpacingBottom] for styles that will cover
+    /// the entire line.
+    max,
+
+    /// Extends the top and bottom edge of the bounds to fully cover any line
+    /// spacing.
+    ///
+    /// The top and bottom of each box will cover half of the
+    /// space above and half of the space below the line. The text should be
+    /// centered vertically within the box.
+    ///
+    /// {@template flutter.dart:ui.boxHeightStyle.includeLineSpacing}
+    /// The top edge of each line should be the same as the bottom edge
+    /// of the line above. There should be no gaps in vertical coverage given any
+    /// amount of line spacing. Line spacing is not included above the first line
+    /// below the last line due to no additional space present there.
+    /// {@endtemplate}
+    includeLineSpacingMiddle,
+
+    /// Extends the top edge of the bounds to fully cover any line spacing.
+    ///
+    /// The line spacing will be added to the top of the box.
+    ///
+    /// {@macro flutter.dart:ui.rectHeightStyle.includeLineSpacing}
+    includeLineSpacingTop,
+
+    /// Extends the bottom edge of the bounds to fully cover any line spacing.
+    ///
+    /// The line spacing will be added to the bottom of the box.
+    ///
+    /// {@macro flutter.dart:ui.boxHeightStyle.includeLineSpacing}
+    includeLineSpacingBottom,
+}
+
+/// Defines various ways to horizontally bound the boxes returned by
+/// [Paragraph.getBoxesForRange].
+enum BoxWidthStyle {
+    // Provide tight bounding boxes that fit widths to the runs of each line
+    // independently.
+    tight,
+
+    /// Adds up to two additional boxes as needed at the beginning and/or end
+    /// of each line so that the widths of the boxes in line are the same width
+    /// as the widest line in the paragraph. The additional boxes are only added
+    /// when the first/last box does not span the max width of the paragraph.
```

Comment:
"The additional boxes on each line are only added when the relevant box at the relevant edge of that line does not span the maximum width of the paragraph."